tooling/nargo_cli/examples/ssa_pass_impact.rs

```diff
@@ -0,0 +1,411 @@
+};
+use noirc_driver::{
+    CompilationResult, CompileOptions, CrateName, NOIR_ARTIFACT_VERSION_STRING, check_crate,
+};
+use noirc_errors::CustomDiagnostic;
+use noirc_evaluator::{
+    brillig::BrilligOptions,
+    errors::RuntimeError,
+    ssa::{SsaEvaluatorOptions, SsaLogging, SsaPass, primary_passes, ssa_gen},
+};
+use noirc_frontend::{
+    debug::DebugInstrumenter,
+    elaborator::UnstableFeature,
+    hir::ParsedFiles,
+    monomorphization::{ast::Program, monomorphize},
+};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use regex::Regex;
+use similar::{ChangeTag, DiffableStr, DiffableStrRef, TextDiff};
+
+/// SSA rendered to `String` after a certain step.
+struct SsaPrint {
+    step: usize,
+    msg: String,
+    ssa: String,
+}
+struct SsaBeforeAndAfter {
+    before: SsaPrint,
+    after: SsaPrint,
+}
+
+/// Try to find the directory that Cargo sets when it is running;
+/// otherwise fallback to assuming the CWD is the root of the repository
+/// and append the crate path.
+fn test_programs_dir() -> PathBuf {
+    let root_dir = match std::env::var("CARGO_MANIFEST_DIR") {
+        Ok(dir) => PathBuf::from(dir).parent().unwrap().parent().unwrap().to_path_buf(),
+        Err(_) => std::env::current_dir().unwrap(),
+    };
+    root_dir.join("test_programs")
+}
+
+/// Collect the test programs under a sub-directory.
+fn read_test_program_dirs(
+    test_programs_dir: &Path,
+    test_sub_dir: &str,
+) -> impl Iterator<Item = PathBuf> + use<> {
+    let test_case_dir = test_programs_dir.join(test_sub_dir);
+    std::fs::read_dir(test_case_dir)
+        .unwrap()
+        .flatten()
+        .filter(|c| c.path().is_dir())
+        .map(|c| c.path())
+}
+
+/// Read a given program directory into a workspace.
+fn read_workspace(
+    program_dir: &Path,
+    selection: PackageSelection,
+) -> Result<Workspace, ManifestError> {
+    let toml_path = get_package_manifest(program_dir)?;
+
+    let workspace = resolve_workspace_from_toml(
+        &toml_path,
+        selection,
+        Some(NOIR_ARTIFACT_VERSION_STRING.to_owned()),
+    )?;
+
+    Ok(workspace)
+}
+
+#[derive(Parser, Debug)]
+struct Options {
+    /// Name of the SSA pass we want to see the impact of.
+    #[arg(long)]
+    ssa_pass: String,
+
+    /// Inliner aggressiveness to use in SSA passes.
+    #[arg(long, default_value = "0")]
+    inliner_aggressiveness: i64,
+
+    /// Show the top N most impacted program passes.
+    #[arg(long, default_value = "50")]
+    top_impact_count: usize,
+}
+
+fn main() {
+    let opts = Options::parse();
+    let sel = PackageSelection::DefaultOrAll;
+
+    let test_workspaces = read_test_program_dirs(&test_programs_dir(), "execution_success")
+        .filter_map(|dir| read_workspace(&dir, sel.clone()).ok())
+        .collect::<Vec<_>>();
+
+    let compile_options = CompileOptions {
+        unstable_features: vec![UnstableFeature::Enums],
```

Comment:
TODO to allow other features?

Reply:
Actually I just I copied this from `test_transform_program_is_idempotent`, which also runs all the integration tests. I don't have any immediately plans to enable any other features, as long as the tests compile. 

But that reminds me to at least print a warning when a program fails to compile, rather than skip it.